File: code-awesome/python-flask-react/todo/backend/src/pages/todo-list.py
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
import os
import sys

# Assuming Flask is already imported in the main script
# This is just to ensure the backend can access it
app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///todo.db'
db = SQLAlchemy(app)

# Define the Todo model
class Todo(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(255))
    description = db.Column(db.Text)

    def __repr__(self):
        return f"<Todo> {self.title} - {self.description}"


# Create the database table if it doesn't exist
def create_db():
    db.create_all()


# API endpoints for the Todo list
@app.route('/todos', methods=['POST'])
def create_todo():
    data = request.get_json()
    try:
        title = data['title']
        description = data['description']
    except KeyError:
        return jsonify({'error': 'Missing title or description'}), 400

    new_todo = Todo(title=title, description=description)
    db.session.add(new_todo)
    db.session.commit()
    return jsonify(new_todo.to_dict()), 201


@app.route('/todos', methods=['GET'])
def get_todos():
    todos = Todo.query.all()
    return jsonify([todo.to_dict() for todo in todos]), 200


@app.route('/todos/<int:id>', methods=['DELETE'])
def delete_todo(id):
    todo = Todo.query.get(id)
    if todo is None:
        return jsonify({'error': 'Todo not found'}), 404
    db.session.delete(todo)
    db.session.commit()
    return jsonify({'message': 'Todo deleted successfully'}), 200


@app.route('/todos/<int:id>', methods=['PUT'])
def update_todo(id):
    todo = Todo.query.get(id)
    if todo is None:
        return jsonify({'error': 'Todo not found'}), 404

    data = request.get_json()
    try:
        todo.title = data['title']
        todo.description = data['description']
    except KeyError:
        return jsonify({'error': 'Missing title or description'}), 400

    db.session.commit()
    return jsonify(todo.to_dict()), 200


if __name__ == '__main__':
    create_db()
    app.run(debug=True)

File: code-awesome/go-gin-react/todo/backend/src/pages/todo-list.go
package main

import (
	"fmt"
	"log"

	"github.com/gin-gonic/gin"
)

type Todo struct {
	ID        int
	Title     string
	Description string
	Completed bool
}

var todos []Todo

func main() {
	router := gin.New()

	router.GET("/", getTodos)
	router.POST("/", createTodo, getTodos)
	router.PUT("/:id", updateTodo, getTodos)
	router.DELETE("/:id", deleteTodo, getTodos)

	router.Run(":8080")
}

func getTodos(c *gin.Context) {
	c.JSON(200, todos)
}

func createTodo(c *gin.Context) {
	todo := Todo{
		Title:     c.PostForm("title"),
		Description: c.PostForm("description"),
		Completed: false,
	}
	todos = append(todos, todo)
	c.JSON(201, todo)
}

func updateTodo(c *gin.Context) {
	id := c.Params.Get("id")
	var todoToUpdate Todo

	for i, t := range todos {
		if fmt.Sprintf("%d", t.ID) == id {
			todoToUpdate = t
			todos[i] = todoToUpdate
			break
		}
	}

	if todoToUpdate.ID == 0 {
		c.Status(404)
		return
	}

	c.JSON(200, todoToUpdate)
}

func deleteTodo(c *gin.Context) {
	id := c.Params.Get("id")
	var todoToDelete Todo

	for i, t := range todos {
		if fmt.Sprintf("%d", t.ID) == id {
			todoToDelete = t
			todos = append(todos[:i], todos[i+1:]...)
			break
		}
	}

	if todoToDelete.ID == 0 {
		c.Status(404)
		return
	}

	c.JSON(200, todoToDelete)
}